scripts/generator/graphql_generator/csharp/Cart.cs.erb

```diff
@@ -46,55 +60,130 @@ namespace <%= namespace %> {
+                    DeletedLineItems.Clear();
+
+                    UpdateCheckout(response.checkoutLineItemsAdd().checkout());
+
+                    CheckoutPoll(callback);
+                }
+            });
+        }
 
-                // convert from variant gid to variant id
-                string[] variandIdSplit = variantId.Split('/');
+        private void UpdateCheckout(Checkout checkout) {
+            if (CurrentCheckout == null) {
+                CurrentCheckout = checkout;
+            } else {
+                MergeCheckout merger = new MergeCheckout();
 
-                url.Append(variandIdSplit[variandIdSplit.Length - 1]);
-                url.Append(":");
-                url.Append(lineItem.Quantity);
+                CurrentCheckout = merger.Merge(CurrentCheckout, checkout);    
             }
 
-            url.Append("?access_token=");
-            url.Append(Client.AccessToken);
+            UpdateLineItemFromCheckout(CurrentCheckout);
+        }
+
+        private void UpdateLineItemFromCheckout(Checkout checkout) {
+            // sometimes we may not query line items for instance when polling is being performed
+            try {
+                List<CheckoutLineItem> lineItems = (List<CheckoutLineItem>) checkout.lineItems();
+
+                LineItems.UpdateLineItemsFromCheckoutLineItems(lineItems);
+            } catch(NoQueryException exception) {}
+        }
+
+        private void CheckoutPoll(Action<List<string>, string> callback) {
+            if (CurrentCheckout.ready()) {
+                callback(null, null);
+            } else {
+                QueryRootQuery query = new QueryRootQuery();
 
-            if (note != null) {
-                url.Append("&note=");
-                url.Append(note);
+                DefaultQueries.checkout.Poll(query, CurrentCheckout.id());
+
+                Client.Query(query, (QueryRoot response, List<string> errors, string httpError) => {
+                    if (errors != null || httpError != null) {
+                        callback(errors, httpError);
+                    } else {
+                        Checkout checkout = (Checkout) response.node();
+
+                        UpdateCheckout(checkout);
+
+                        if (CurrentCheckout.ready()) {
+                            callback(null, null);
+                        } else {
+                            #if !SHOPIFY_MONO_UNIT_TEST
+                                UnityTimeout.Start(POLL_DELAY_SECONDS, () => {
+                                    CheckoutPoll(callback);    
+                                });
+                            #else
+                                CheckoutPoll(callback);
+                            #endif
+                        }
+                    }
+                });
             }
+        }
 
-            return url.ToString();
+        private void OnDeleteLineItem(string lineItemId) {
+            DeletedLineItems.Add(lineItemId);
         }
     }
 }
```

Comment:
Maybe like a Purchase object that encapsulates the high level operations happening on cart and checkout? we can think about this for a refactor if you want instead of addressing it here, but since this does change the interface I would say that its more favorable to address it before shipping this pr...